OctoAwesome/OctoAwesome/Components/ComponentList.cs

```diff
@@ -9,202 +8,360 @@
         {
-            if (!components.TryGetValue(typeof(V), out var component))
-                return false;
+            onRemover?.Invoke(component);
 
-            removeValidator?.Invoke(component);
-            if (components.Remove(typeof(V)))
-            {
-                onRemover?.Invoke(component);
-                return true;
-            }
+            if (componentsByType.TryGetValue(typeof(V), out var components))
+                components.Remove(component);
+
+            return true;
+        }
+
+        return false;
+    }
+
+    /// <summary>
+    /// Replaces one component with another and retrieves the replaces value when any was found
+    /// </summary>
+    /// <typeparam name="V">The type to to replace</typeparam>
+    /// <param name="toReplace">Component to search for to be replaced</param>
+    /// <param name="replacement">Component that replaces the <paramref name="toReplace"/></param>
+    /// <param name="replaced">The item that equaled <paramref name="toReplace"/> and was replaced</param>
+    /// <returns><see langword="true"/> if <paramref name="toReplace"/> was found, otherwise <see langword="false"/></returns>
+    public virtual bool Replace<V>(V toReplace, V replacement, [MaybeNullWhen(false)] out V replaced) where V : T
+    {
 
+        if (!componentsByType.TryGetValue(typeof(V), out var components))
+        {
+            replaced = default;
             return false;
         }
 
-        /// <summary>
-        /// Serializes the component list to a binary writer..
-        /// </summary>
-        /// <param name="writer">The binary writer to serialize the component list to.</param>
-        public virtual void Serialize(BinaryWriter writer)
+        var index = components.IndexOf(toReplace);
+        if (index < 0)
         {
-            writer.Write(components.Count);
-            foreach (var componente in components)
-            {
-                writer.Write(componente.Key.AssemblyQualifiedName!);
-                componente.Value.Serialize(writer);
+            replaced = default;
+            return false;
+        }
+
+        replaced = GenericCaster<T, V>.Cast(components[index])!;
+        components[index] = replacement;
+        return true;
+    }
+
+    /// <summary>
+    /// Replaces <paramref name="toReplace"/> with replacements or just adds <paramref name="replacement"/>, if <paramref name="toReplace"/> was not found
+    /// </summary>
+    /// <typeparam name="V">Type to add or replace</typeparam>
+    /// <param name="toReplace">Component to search for to be replaced</param>
+    /// <param name="replacement">Component that replaces the <paramref name="toReplace"/></param>
+    /// <param name="replaced">The item that equaled <paramref name="toReplace"/> and was replaced</param>
+    /// <returns><see langword="false"/> if only insert took place, otherwise <see langword="true"/> when <paramref name="toReplace"/> was found and replaced</returns>
```

Comment:
```suggestion
    /// Replace <paramref name="toReplace"/> with <paramref name="replacement"/> or just add <paramref name="replacement"/>, if <paramref name="toReplace"/> was not found.
    /// </summary>
    /// <typeparam name="V">The component type to add or replace</typeparam>
    /// <param name="toReplace">The component to search for to be replaced.</param>
    /// <param name="replacement">The component that replaces <paramref name="toReplace"/>.</param>
    /// <param name="replaced">The item that was equal to <paramref name="toReplace"/> and was replaced.</param>
    /// <returns><see langword="false"/> if only insert took place, otherwise <see langword="true"/> when <paramref name="toReplace"/> was found and replaced</returns>
```